tests/flytekit/unit/core/test_type_hints.py

```diff
@@ -323,13 +323,30 @@ def t2(a: typing.List[int], b: typing.List[str]) -> (int, str):
 
     @workflow
     def my_wf(a: typing.List[int]) -> (int, str):
-        x, y = maptask(t1, metadata=TaskMetadata(retries=1))(a=a)
+        x, y = map(t1, metadata=TaskMetadata(retries=1))(a=a)
         return t2(a=x, b=y)
 
     x = my_wf(a=[5, 6])
     assert x == (15, "world-7world-8")
 
 
+def test_map_task():
```

Comment:
Can we add the following tests?

* some [serialization test](https://github.com/flyteorg/flytekit/blob/master/tests/flytekit/unit/core/test_serialization.py#L15) something where we serialize a task, and look at it to make sure the task type is correct.
* what does serialization look like if the user does
```
mapped_t1 = map(t1, metadata=TaskMetadata(retries=1))

@workflow
def wf1():
    return mapped_t1(a=a)

@workflow
def wf2():
    map(t1, metadata=TaskMetadata(retries=1))(a=a)
```
there's basically two map tasks at this point right?  coming from the same `t1`.  At serialization, will this produce two protobufs?  Will they collide?  Are the names unique?
* a serialization test for the workflow
* test local workflow execution
* also what happens if you map over a launch plan or a sub wf?  should those error? 
* what happens if you map over a task that has multiple inputs?
* what happens if you map over a task that has multiple outputs?

Also can you remind me what restrictions there are here?  Can users do this?
```
@workflow
def wf() -> (something, smth_else):
  results = map(t1, metadata=TaskMetadata(retries=1))(a=a)
  x = t2(list_of_ints=results[0:5])
  y = t2(list_of_ints=results[5:])
  return x, y
```
no right?  What about in a dynamic task?

If we have a dynamic task that `range`s over something, and there ends up being N copies of the same task, does the current implementation preclude future optimization of those N copies into one array task?

Reply:
note this test case already tests `test local workflow execution`

Reply:
>  Can users do this?

No this should be unchanged because the results are still a promise.

To answer your last question, we can certainly update dynamic tasks to produce array jobs (that was my initial approach). It's just mildly tedious to implement but certainly doable especially given that the old sdk just does this.

Added test cases for everything else above.